### DISCUSSION TITLE
projectile-find-file fuzzy matching not working for me in projectile 2.3.0 on mac

### DISCUSSION THREAD

**Discussion by steinarb** [bbatsov/projectile#1646] [Q&A] [answered]
Hi,

In projectile 2.0.0, emacs-26.1 on debian 10.7 "buster" and in projectile 2.1.0, emacs 26.3 on ubuntu 18.04, fuzzy matching of the filename in projectile-find-file works.

I.e. I can start typing the file name and projectile shows the alternatives, and narrows down the alternatives as I type.

However, in projectile 2.3.0, emacs 27.1 on mac OS/X fuzzy matching of the file name doesn't work. I have to type in the full path of the file starting at the top of the path.

I'm trying to figure out if this is caused by:

1. A change in projectile between 2.1.0 and 2.3.0? (I don't think so, because the animated demo in the github README still shows the behaviour I want, I can't see anything relevant in the changelog and google hasn't found anything about this)
2. A difference between emacs 26.x and emacs 27.x?
3. A difference between emacs on GNU/linux and mac OS/X?
4. A difference in config? I have only this config on all three:
    ```
    (projectile-mode +1)
    (define-key projectile-mode-map (kbd "s-p") 'projectile-command-map)
    (define-key projectile-mode-map (kbd "C-c p") 'projectile-command-map)
    (setq projectile-enable-caching 't)
    ```

I don't use helm on any of the 3 systems. Use of helm-projectile came up in some of the google searches.

Does anyone know what I might try to get the fuzzy file name matching working on the mac?

Thanks!

- Steinar

**Comment by bbatsov** [comment, marked as answer]
The fuzzy matching doesn't come from Projectile, but from whatever minibuffer selection framework your're using (E.g. Helm, Ivy, Selectrum). Projectile just returns a list of filenames and it's up to the other tools to figure out what to do with them.

**Reply by steinarb** [reply]
Thanks!

Is there any way of figuring out which one(s) I'm using on debian and/or ubuntu?

I'm pretty sure it isn't helm.

**Reply by steinarb** [reply]
The value of projectile-completion-system seems to be a clue.

On the debian box, at least, it is:
```
projectile-completion-system is a variable defined in ‘projectile.el’.
Its value is ‘ido’
```

**Reply by bbatsov** [reply]
If you're using `ido` you should also install `ido-flx` for fuzzy matching. Personally, I like selectrum the most.

**Reply by steinarb** [reply]
I now know how I installed projectile on debian and where flx-ido came from: I installed projectile from [a debian package](https://packages.debian.org/buster/elpa-projectile) which in turn pulled in flx-ido from [another debian package](https://packages.debian.org/buster/elpa-flx-ido).

I.e. 
```
apt install elpa-projectile
```

Presumably I repeated that install on ubuntu (I don't have access to the ubuntu machine right now).

**Reply by steinarb** [reply]
Ok, this was kind of interesting.  

I now have access to both the Ubuntu machine and the mac, and this is what I've seen and done:

1. The ubuntu machine had `projectile` installed directly from melpa stable, rather than via a debian package
2. The value of `projectile-completion-system` was `ido`
3. The ubuntu machine did _not_ have `flx-ido` installed
4. But still fuzzy-completion worked in projectile on the ubuntu machine (and I couldn't figure out why...?)
5. Back on the mac, I tried installing `selectrum` from melpa stable (since that was the completion system recommended by @bbatsov )
6. The value of `projectile-completion-system` on the mac, was `auto`
7. However, after installing `selectrum`, fuzzy completion still didn't work in projectile-find-file
8. So I googled and found this issue: https://github.com/raxod502/selectrum/issues/112
9. I recognized the completion style in the screenshot, because that's what it looks like on the ubuntu machine
10. A [comment in the issue](https://github.com/raxod502/selectrum/issues/112#issuecomment-644276120) says: That looks like you use icomplete?
11. I did a `locate-library` for icomplete and found that this was something delivered with emacs
12. So I customized package-completion-system and set the value to `ido` (even though I didn't have flx-ido installed)
13. I now had icompletion-style fuzzy matching on the mac
14. I tried switching package-completion-system to `default`(which is the other value mentioned in the issue above), but that gave me no fuzzy completion
15. Setting the value of package-completion-system back to `ido` gave me icompletion style fuzzy matching back

So in conclusion:

1. I now have fuzzy completion in projectile, but not the recommended completion, and by using a confusing configuration setting
2. I still haven't been able to make selectrum work (but I can live with icomplete for now)

thanks!

- Steinar
